File: internal/plugins/jest/rules/no_standalone_expect/no_standalone_expect_test.go
package no_standalone_expect_test

import (
	"testing"

	"github.com/web-infra-dev/rslint/internal/plugins/jest/fixtures"
	"github.com/web-infra-dev/rslint/internal/plugins/jest/rules/no_standalone_expect"
	"github.com/web-infra-dev/rslint/internal/rule_tester"
)

func TestNoStandaloneExpectRule(t *testing.T) {
	rule_tester.RunRuleTester(
		fixtures.GetRootDir(),
		"tsconfig.json",
		t,
		&no_standalone_expect.NoStandaloneExpectRule,
		[]rule_tester.ValidTestCase{
			{Code: `expect.any(String)`},
			{Code: `expect.extend({})`},
			{Code: `expect.not.stringContaining("value")`},
			{Code: `describe("a test", () => { it("an it", () => {expect(1).toBe(1); }); });`},
			{Code: `describe("a test", () => { it("an it", () => { const func = () => { expect(1).toBe(1); }; }); });`},
			{Code: `describe("a test", () => { const func = () => { expect(1).toBe(1); }; });`},
			{Code: `describe("a test", () => { function func() { expect(1).toBe(1); }; });`},
			{Code: `describe("a test", () => { const func = function(){ expect(1).toBe(1); }; });`},
			{Code: `it("an it", () => expect(1).toBe(1))`},
			{Code: `const func = function(){ expect(1).toBe(1); };`},
			{Code: `const func = () => expect(1).toBe(1);`},
			{Code: `{}`},
			{Code: `it.each([1, true])("trues", value => { expect(value).toBe(true); });`},
			{Code: `it.each([1, true])("trues", value => { expect(value).toBe(true); }); it("an it", () => { expect(1).toBe(1) });`},
			{Code: `
      it.each` + "`" + `
        num   | value
        ${1} | ${true}
      ` + "`" + `('trues', ({ value }) => {
        expect(value).toBe(true);
      });
    `},
			{Code: `it.only("an only", value => { expect(value).toBe(true); });`},
			{Code: `it.concurrent("an concurrent", value => { expect(value).toBe(true); });`},
			{Code: `describe.each([1, true])("trues", value => { it("an it", () => expect(value).toBe(true) ); });`},
			{Code: `const helpers = { assert() { expect(1).toBe(1); } };`},
			{Code: `class Helper { assert() { expect(1).toBe(1); } get value() { expect(1).toBe(1); return 1; } }`},
			{
				Code: `
        describe('scenario', () => {
          const t = Math.random() ? it.only : it;
          t('testing', () => expect(true));
        });
      `,
				Options: []interface{}{
					map[string]interface{}{"additionalTestBlockFunctions": []interface{}{"t"}},
				},
			},
			{
				Code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
				Options: []interface{}{
					map[string]interface{}{"additionalTestBlockFunctions": []interface{}{"each.test"}},
				},
			},
		},
		[]rule_tester.InvalidTestCase{
			{
				Code: "(() => {})('testing', () => expect(true).toBe(false))",
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 29, EndColumn: 53},
				},
			},
			{
				Code: `expect.hasAssertions()`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 1, EndColumn: 23},
				},
			},
			{
				Code: `expect().hasAssertions()`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 1, EndColumn: 25},
				},
			},
			{
				Code: `
        describe('scenario', () => {
          const t = Math.random() ? it.only : it;
          t('testing', () => expect(true).toBe(false));
        });
      `,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 4, Column: 30, EndColumn: 54},
				},
			},
			{
				Code: `
        describe('scenario', () => {
          const t = Math.random() ? it.only : it;
          t('testing', () => expect(true).toBe(false));
        });
      `,
				Options: []interface{}{
					map[string]interface{}{"additionalTestBlockFunctions": nil},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 4, Column: 30, EndColumn: 54},
				},
			},
			{
				Code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 7, Column: 11, EndColumn: 39},
				},
			},
			{
				Code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
				Options: []interface{}{
					map[string]interface{}{"additionalTestBlockFunctions": []interface{}{"each"}},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 7, Column: 11, EndColumn: 39},
				},
			},
			{
				Code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
				Options: []interface{}{
					map[string]interface{}{"additionalTestBlockFunctions": []interface{}{"test"}},
				},
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 7, Column: 11, EndColumn: 39},
				},
			},
			{
				Code: `describe("a test", () => { expect(1).toBe(1); });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 28, EndColumn: 45},
				},
			},
			{
				Code: `describe("a test", () => expect(1).toBe(1));`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 26, EndColumn: 43},
				},
			},
			{
				Code: `describe("a test", () => { const func = () => { expect(1).toBe(1); }; expect(1).toBe(1); });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 71, EndColumn: 88},
				},
			},
			{
				Code: `describe("a test", () => {  it(() => { expect(1).toBe(1); }); expect(1).toBe(1); });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 63, EndColumn: 80},
				},
			},
			{
				Code: `expect(1).toBe(1);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 1, EndColumn: 18},
				},
			},
			{
				Code: `{expect(1).toBe(1)}`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 2, EndColumn: 19},
				},
			},
			{
				Code: `it.each([1, true])("trues", value => { expect(value).toBe(true); }); expect(1).toBe(1);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 70, EndColumn: 87},
				},
			},
			{
				Code: `it.only("an only", () => { expect(1).toBe(1); }); expect(2).toBe(2);`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 51, EndColumn: 68},
				},
			},
			{
				Code: `describe.each([1, true])("trues", value => { expect(value).toBe(true); });`,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 1, Column: 46, EndColumn: 70},
				},
			},
			{
				Code: `
        describe.each` + "`" + `
          num   | value
          ${1} | ${true}
        ` + "`" + `('trues', ({ value }) => {
          expect(value).toBe(true);
        });
      `,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 6, Column: 11, EndColumn: 35},
				},
			},
			{
				Code: `
        it.each` + "`" + `
          num   | value
          ${1} | ${true}
        ` + "`" + `('trues', ({ value }) => {
          expect(value).toBe(true);
        });

        expect(1).toBe(1);
      `,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 9, Column: 9, EndColumn: 26},
				},
			},
			{
				Code: `
        import { expect as pleaseExpect } from '@jest/globals';

        describe("a test", () => { pleaseExpect(1).toBe(1); });
      `,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 4, Column: 36, EndColumn: 59},
				},
			},
			{
				Code: `
        import { expect as pleaseExpect } from '@jest/globals';

        beforeEach(() => pleaseExpect.hasAssertions());
      `,
				Errors: []rule_tester.InvalidTestCaseError{
					{MessageId: "unexpectedExpect", Line: 4, Column: 26, EndColumn: 54},
				},
			},
		},
	)
}

File: packages/rslint-test-tools/tests/eslint-plugin-jest/rules/no-standalone-expect.test.ts
import { RuleTester } from '../rule-tester';

const ruleTester = new RuleTester();

ruleTester.run('no-standalone-expect', {} as never, {
  valid: [
    { code: 'expect.any(String)' },
    { code: 'expect.extend({})' },
    {
      code: 'describe("a test", () => { it("an it", () => {expect(1).toBe(1); }); });',
    },
    {
      code: 'describe("a test", () => { it("an it", () => { const func = () => { expect(1).toBe(1); }; }); });',
    },
    {
      code: 'describe("a test", () => { const func = () => { expect(1).toBe(1); }; });',
    },
    {
      code: 'describe("a test", () => { function func() { expect(1).toBe(1); }; });',
    },
    {
      code: 'describe("a test", () => { const func = function(){ expect(1).toBe(1); }; });',
    },
    { code: 'it("an it", () => expect(1).toBe(1))' },
    { code: 'const func = function(){ expect(1).toBe(1); };' },
    { code: 'const func = () => expect(1).toBe(1);' },
    { code: '{}' },
    {
      code: 'it.each([1, true])("trues", value => { expect(value).toBe(true); });',
    },
    {
      code: 'it.each([1, true])("trues", value => { expect(value).toBe(true); }); it("an it", () => { expect(1).toBe(1) });',
    },
    {
      code: `
      it.each\`
        num   | value
        \${1} | \${true}
      \`('trues', ({ value }) => {
        expect(value).toBe(true);
      });
    `,
    },
    { code: 'it.only("an only", value => { expect(value).toBe(true); });' },
    {
      code: 'it.concurrent("an concurrent", value => { expect(value).toBe(true); });',
    },
    {
      code: 'describe.each([1, true])("trues", value => { it("an it", () => expect(value).toBe(true) ); });',
    },
    {
      code: 'describe.each([1, true])("trues", value => { it("an it", () => expect(value).toBe(true) ); });',
    },
    {
      code: 'const helpers = { assert() { expect(1).toBe(1); } };',
    },
    {
      code: 'class Helper { assert() { expect(1).toBe(1); } get value() { expect(1).toBe(1); return 1; } }',
    },
    {
      code: `
        describe('scenario', () => {
          const t = Math.random() ? it.only : it;
          t('testing', () => expect(true));
        });
      `,
      options: { additionalTestBlockFunctions: ['t'] },
    },
    {
      code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
      options: { additionalTestBlockFunctions: ['each.test'] },
    },
  ],
  invalid: [
    {
      code: "(() => {})('testing', () => expect(true).toBe(false))",
      errors: [{ endColumn: 53, column: 29, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'expect.hasAssertions()',
      errors: [{ endColumn: 23, column: 1, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'expect().hasAssertions()',
      errors: [{ endColumn: 25, column: 1, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        describe('scenario', () => {
          const t = Math.random() ? it.only : it;
          t('testing', () => expect(true).toBe(false));
        });
      `,
      errors: [{ endColumn: 46, column: 22, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        describe('scenario', () => {
          const t = Math.random() ? it.only : it;
          t('testing', () => expect(true).toBe(false));
        });
      `,
      options: { additionalTestBlockFunctions: undefined },
      errors: [{ endColumn: 46, column: 22, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
      errors: [{ endColumn: 31, column: 3, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
      options: { additionalTestBlockFunctions: ['each'] },
      errors: [{ endColumn: 31, column: 3, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        each([
          [1, 1, 2],
          [1, 2, 3],
          [2, 1, 3],
        ]).test('returns the result of adding %d to %d', (a, b, expected) => {
          expect(a + b).toBe(expected);
        });
      `,
      options: { additionalTestBlockFunctions: ['test'] },
      errors: [{ endColumn: 31, column: 3, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'describe("a test", () => { expect(1).toBe(1); });',
      errors: [{ endColumn: 45, column: 28, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'describe("a test", () => expect(1).toBe(1));',
      errors: [{ endColumn: 43, column: 26, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'describe("a test", () => { const func = () => { expect(1).toBe(1); }; expect(1).toBe(1); });',
      errors: [{ endColumn: 88, column: 71, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'describe("a test", () => {  it(() => { expect(1).toBe(1); }); expect(1).toBe(1); });',
      errors: [{ endColumn: 80, column: 63, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'expect(1).toBe(1);',
      errors: [{ endColumn: 18, column: 1, messageId: 'unexpectedExpect' }],
    },
    {
      code: '{expect(1).toBe(1)}',
      errors: [{ endColumn: 19, column: 2, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'it.each([1, true])("trues", value => { expect(value).toBe(true); }); expect(1).toBe(1);',
      errors: [{ endColumn: 87, column: 70, messageId: 'unexpectedExpect' }],
    },
    {
      code: 'describe.each([1, true])("trues", value => { expect(value).toBe(true); });',
      errors: [{ endColumn: 70, column: 46, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        describe.each\`
          num   | value
          \${1} | \${true}
        \`('trues', ({ value }) => {
          expect(value).toBe(true);
        });
      `,
      errors: [{ endColumn: 27, column: 3, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        it.each\`
          num   | value
          \${1} | \${true}
        \`('trues', ({ value }) => {
          expect(value).toBe(true);
        });

        expect(1).toBe(1);
      `,
      errors: [{ endColumn: 17, column: 1, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        import { expect as pleaseExpect } from '@jest/globals';

        describe("a test", () => { pleaseExpect(1).toBe(1); });
      `,
      errors: [{ endColumn: 51, column: 28, messageId: 'unexpectedExpect' }],
    },
    {
      code: `
        import { expect as pleaseExpect } from '@jest/globals';

        beforeEach(() => pleaseExpect.hasAssertions());
      `,
      errors: [{ endColumn: 46, column: 18, messageId: 'unexpectedExpect' }],
    },
  ],
});
